core/src/main/java/org/elasticsearch/index/query/NestedQueryBuilder.java

```diff
@@ -88,4 +138,89 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
     public final String getWriteableName() {
         return NAME;
     }
+
+    @Override
+    protected boolean doEquals(NestedQueryBuilder that) {
+        return Objects.equals(query, that.query)
+                && Objects.equals(path, that.path)
+                && Objects.equals(scoreMode, that.scoreMode)
+                && Objects.equals(queryInnerHits, that.queryInnerHits);
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(query, path, scoreMode, queryInnerHits);
+    }
+
+    public NestedQueryBuilder(StreamInput in) throws IOException {
+        path = in.readString();
+        final int ordinal = in.readVInt();
+        scoreMode = ScoreMode.values()[ordinal];
+        query = in.readQuery();
+        if (in.readBoolean()) {
+            queryInnerHits = new QueryInnerHits(in);
+        }
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(path);
+        out.writeVInt(scoreMode.ordinal());
+        out.writeQuery(query);
+        if (queryInnerHits != null) {
+            out.writeBoolean(true);
+            queryInnerHits.writeTo(out);
+        } else {
+            out.writeBoolean(false);
+        }
+    }
+
+    @Override
+    protected NestedQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        return new NestedQueryBuilder(in);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        ObjectMapper nestedObjectMapper = context.getObjectMapper(path);
+        final BitSetProducer parentFilter;
```

Comment:
shouldn't we check that the returned mapper is not null and is od nested type? we have this in NestedInnerQueryParseSupport: 

```
if (nestedObjectMapper == null) {
            throw new QueryParsingException(parseContext, "[nested] failed to find nested object under path [" + path + "]");
        }
        if (!nestedObjectMapper.nested().isNested()) {
            throw new QueryParsingException(parseContext, "[nested] nested object under path [" + path + "] is not of nested type");
        }
```